### DISCUSSION TITLE
Missing Letters

### DISCUSSION THREAD

**Discussion by Thomas-Lerman** [pelicanmapping/osgearth#2318] [Support Q&A] [answered]
We have run into this in the past and I never thought to ask about it. On NVIDIA, everything looks good. Other graphics cards, especially Intel (even their "high performance" cards), we see some deviations. The most common is missing text. Please see attached image. In this case, we have realized that the numbers, the lower-case letters have the same working letters (the place mark with no text are the non-working letters), and the upper-case letters have a different set of working letters. Anyone seem this or have any ideas? The font is LiberationSans-Regular.ttf. As mentioned, the same thing on NVIDIA looks good.
![image](https://github.com/gwaldron/osgearth/assets/1148649/7a0a52bd-c855-46c9-94bb-42ea84a1327d)

**Comment by emminizer** [comment, marked as answer]
This looks very similar to an issue we saw in 2015. One suggestion we had was to try to set environment variable `OSG_TEXT_INCREMENTAL_SUBLOADING=OFF` but my notes indicate that did not help. In fact once I found the solution IIRC this got in the way.

Another place I found indicated that the glyph margins were a big impact on this -- `Font::setGlyphImageMargin(2)`. But osgEarth should be doing this automatically.

I had a post on the old OSG forums about it, but it looks like it was nuked -- http://forum.openscenegraph.org/viewtopic.php?p=65332 was the link.

It looks like I came across https://stackoverflow.com/questions/30389675/how-to-debug-opengl-a-gray-black-texture-box which pointed me to the solution. This was with an older version of OSG -- I believe the code is gone in 3.6.5; what OSG version are you using? My patch on OSG that fixed the issue was this:

```
diff -r a80d8687e9be src/osgText/Glyph.cpp
--- a/src/osgText/Glyph.cpp	Fri Jan 29 07:24:45 2016 -0500
+++ b/src/osgText/Glyph.cpp	Fri Jan 29 07:47:43 2016 -0500
@@ -373,6 +373,7 @@
             glyphsWereSubloading.clear();
 
             glPixelStorei(GL_UNPACK_ALIGNMENT,1);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, getFilter(osg::Texture::MIN_FILTER));
 
             #if !defined(OSG_GLES1_AVAILABLE) && !defined(OSG_GLES2_AVAILABLE)
             glPixelStorei(GL_UNPACK_ROW_LENGTH,getTextureWidth());
@@ -517,6 +518,10 @@
     #if !defined(OSG_GLES1_AVAILABLE) && !defined(OSG_GLES2_AVAILABLE)
     glPixelStorei(GL_UNPACK_ROW_LENGTH,getRowLength());
     #endif
+    if (_texture)
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, _texture->getFilter(osg::Texture::MIN_FILTER));
+    else
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
 
     glTexSubImage2D(GL_TEXTURE_2D,0,
                     _texturePosX,_texturePosY,
```


Maybe this helps. Maybe not. But I know I've seen similar in the past but it's been almost a decade since we saw it.

**Comment by Thomas-Lerman** [comment]
Older version of everything unfortunately. We are somewhat stuff due to some other libraries we are using.

**Reply by emminizer** [reply]
Great -- that might mean that the patch above on OSG might help fix the problem.

**Comment by Thomas-Lerman** [comment]
As expected, the environment variable did not change it. Trying the change suggested hoping I have the additions in the right places. The lines are closer together for me (only see one instance of GL_UNPACK_ROW_LENGTH). Ooops, getFilter() in undefined . . . it seems to be need _texture->getFilter(). Initial tests: no change on NVIDIA and it fixed it on one one computer with an Intel card. I am glad I jumped in here and you were able to respond so quickly. Thank you very much!!!

**Reply by emminizer** [reply]
Awesome! Glad I could help. It took us a LONG time to track down and finally fix that issue. Most of our hardware is NVIDIA too.

**Comment by Thomas-Lerman** [comment]
Marking as answered, tried on two computers that it fixed.

**Comment by Thomas-Lerman** [comment]
In case anyone rans into this, the above fix does not work for Windows 11. The following does and so far seems backwards compatible as well.
```
	glPixelStorei(GL_UNPACK_ALIGNMENT,getPacking());
#ifdef OLD_FIX
    // https://github.com/gwaldron/osgearth/discussions/2318
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, _texture->getFilter(osg::Texture::MIN_FILTER));
#else
    // Intel iGPU fix: force non-mipmapped filtering and clamp mip levels
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
#ifndef GL_TEXTURE_BASE_LEVEL
#  define GL_TEXTURE_BASE_LEVEL 0x813C
#endif // GL_TEXTURE_BASE_LEVEL
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
#ifndef GL_TEXTURE_MAX_LEVEL
#  define GL_TEXTURE_MAX_LEVEL 0x813D
#endif // GL_TEXTURE_MAX_LEVEL
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 0);
#endif // OLD_FIX

    #if !defined(OSG_GLES1_AVAILABLE) && !defined(OSG_GLES2_AVAILABLE)
    glPixelStorei(GL_UNPACK_ROW_LENGTH,getRowLength());
    #endif
#ifdef OLD_FIX
    // https://github.com/gwaldron/osgearth/discussions/2318
    if (_texture)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, _texture->getFilter(osg::Texture::MIN_FILTER));
    else
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
#else
    // (previous conditional MIN filter removed to avoid mipmapped modes)
#endif // OLD_FIX

```
